### PR TITLE
Add Multi-Process Sharding Support

### PR DESCRIPTION
**Pull request by VelvetThePanda**
# Summary
This should make multi-process sharding just that much easier. 

# Details
Adds a new client and configuration option to support multi-process sharding

# Changes proposed
- Unseal DiscordShardedClient
- Add some `private protected` access modifiers where needed
- Add a new client type (DiscordClusterClient)
- Changed ShardedLoggerFactory to be generic (to support correct contextual logging)
- Added ClusterId and ShardBootCount properties to DiscordConfiguration

# Notes
How have we gone over half a decade without this being a thing?

### CODE REVIEW

**Comment by VelvetThePanda** [comment]
I can now say I've thoroughly tested the things I've added. If you spot something wrong do let me know. I wrote tests, those tests passed. 
image


## Test source:
```cs
using System;
using System.Threading.Tasks;
using DSharpPlus;
using NUnit.Framework;

namespace DSharpPlusTest
{
	public sealed class ShardTests
	{
		private const string Token = "Your token here";
		private readonly DiscordConfiguration _noBootShardConfig = new DiscordConfiguration()
		{
			Token = Token,
			ShardCount = 2,
		};
		
		private readonly DiscordConfiguration _bootShardConfig = new DiscordConfiguration()
		{
			Token = Token,
			ShardCount = 2,
			BootShardCount = 2
		};

		private readonly DiscordConfiguration _misMatchedBootConfig = new DiscordConfiguration()
		{
			Token = Token,
			ShardCount = 2,
			BootShardCount = 1
		};
		
		[Test]
		public async Task ShardedClient_Connects_Normally_With_No_BootCount()
		{
			//Arrange
			var client = new DiscordShardedClient(_noBootShardConfig);
			
			//Act
			Assert.DoesNotThrowAsync(async () => await client.StartAsync());
			
			//Assert
			Assert.AreEqual(2,client.ShardClients.Count);
		}

		[Test]
		public async Task ShardedClient_Connects_Normally_With_BootCount()
		{
			//Arrange
			var client = new DiscordShardedClient(_bootShardConfig);
			
			//Act
			Assert.DoesNotThrowAsync(async () => await client.StartAsync());
			
			//Assert
			Assert.AreEqual(2,client.ShardClients.Count);
		}

		[Test]
		public async Task ShardedClient_Connects_Normally_With_SpecifiedBootCount_When_LessThan_ShardCount()
		{
			//Arrange
			var client = new DiscordShardedClient(_misMatchedBootConfig);
			
			//Act
			Assert.DoesNotThrowAsync(async () => await client.StartAsync());
			
			//Assert
			Assert.AreEqual(1,client.ShardClients.Count);
		}

		[Test]
		public async Task ShardedClient_Connects_With_Specified_ShardId_When_Specified()
		{
			//Arrange
			var conf1 = new DiscordConfiguration(_bootShardConfig) {ShardCount = 4, ShardId = 0};
			var conf2 = new DiscordConfiguration(conf1) {ShardId = 2};

			var client1 = new DiscordShardedClient(conf1);
			var client2 = new DiscordShardedClient(conf2);
			
			//Act
			Assert.DoesNotThrowAsync(async () => await client1.StartAsync());
			Assert.DoesNotThrowAsync(async () => await client2.StartAsync());
			
			//Assert
			Assert.AreEqual(0, client1.ShardClients[0].ShardId);
			Assert.AreEqual(2, client2.ShardClients[0].ShardId);
		}
		
		[Test]
		public async Task ShardedClient_OnlyConnects_Specified_Shards()
		{
			//Arrange
			var conf1 = new DiscordConfiguration(_bootShardConfig) {ShardCount = 4, ShardId = 0};
			var conf2 = new DiscordConfiguration(conf1) {ShardId = 2};

			var client1 = new DiscordShardedClient(conf1);
			var client2 = new DiscordShardedClient(conf2);
			
			//Act
			Assert.DoesNotThrowAsync(async () => await client1.StartAsync());
			Assert.DoesNotThrowAsync(async () => await client2.StartAsync());
			
			//Assert
			Assert.AreEqual(2, client1.ShardClients.Count);
			Assert.AreEqual(2, client2.ShardClients.Count);
		}

		[Test]
		public void DiscordConfiguration_Disallows_Invalid_NonNull_Values()
		{
			//Arrange
			var config = new DiscordConfiguration();
			
			//Act
			//Assert
			Assert.Throws<ArgumentOutOfRangeException>(() => config.BootShardCount = 0);
			Assert.Throws<ArgumentOutOfRangeException>(() => config.BootShardCount = 2);
		}
	}
}
```

**Comment by Neuheit** [comment]
I'm still trying to understand the use case of `ShardBootCount`. `ShardCount` seems to do exactly the same thing.

**Comment by VelvetThePanda** [comment]
It does not. ShardCount simply tells Discord how many shards will be connecting to the gateway, and current behavior shows that even if you set ShardId, the DiscordShardedClient will still attempt to start all shards starting from 0, which essentially just connects to the same shards that were connected elsewhere, which is not good.

**Comment by uwx** [comment]
> current behavior shows that even if you set ShardId, the DiscordShardedClient will still attempt to start all shards starting from 0, which essentially just connects to the same shards that were connected elsewhere, which is not good.

Current behavior is correct. A single DiscordShardedClient manages all the shards. There should not be "shards that were connected elsewhere", that is what manual sharding is for.

I highly advise against merging this PR. It is misusing the entire concept of the DiscordShardedClient.

> This allows limiting of how many shards are booted, and for ShardId to actually mean something (since it would just get reset previously.)

It does exactly what it's supposed to do. ShardId is set in the individual clients stored in the DiscordShardedClient, to the index of the client. It's not meant to be assigned in the configuration in this scenario.

> Adds a new property to the DiscordConfiguration that allows configuring of how many shards to boot.

What the fuck?

> How have we gone over half a decade without this being fixed?

Because it isn't a bug.

**Comment by VelvetThePanda** [comment]
After some back and forth, this PR will be restructured. A new client will be formed that takes advantage of the new property and will be dedicated to multi-process sharding.

**Comment by VelvetThePanda** [comment]
new tests are working, so, I'm gonna hope that means this PR is OK?

**Comment by VelvetThePanda** [comment]
Maybe someone has something to say about these tests, but eh.



## ClusterClient test source:

```cs
using System;
using System.Threading.Tasks;
using DSharpPlus;
using NUnit.Framework;

namespace DSharpPlusTest
{
	public sealed class ShardTests
	{
		private const string Token = "Token here :)";
		private readonly DiscordConfiguration _noBootShardConfig = new DiscordConfiguration()
		{
			Token = Token,
			ShardCount = 2,
		};
		
		private readonly DiscordConfiguration _bootShardConfig = new DiscordConfiguration()
		{
			Token = Token,
			ShardCount = 2,
			BootShardCount = 2
		};

		private readonly DiscordConfiguration _misMatchedBootConfig = new DiscordConfiguration()
		{
			Token = Token,
			ShardCount = 2,
			BootShardCount = 1
		};
		
		[Test]
		public async Task ClusterClient_Connects_Normally_With_No_BootCount()
		{
			//Arrange
			var client = new DiscordClusterClient(_noBootShardConfig);
			
			//Act
			Assert.DoesNotThrowAsync(async () => await client.StartAsync());
			
			//Assert
			Assert.AreEqual(2,client.ShardClients.Count);
		}

		[Test]
		public async Task ClusterClient_Connects_Normally_With_BootCount()
		{
			//Arrange
			var client = new DiscordClusterClient(_bootShardConfig);
			
			//Act
			Assert.DoesNotThrowAsync(async () => await client.StartAsync());
			
			//Assert
			Assert.AreEqual(2,client.ShardClients.Count);
		}

		[Test]
		public async Task ClusterClient_Connects_Normally_With_SpecifiedBootCount_When_LessThan_ShardCount()
		{
			//Arrange
			var client = new DiscordClusterClient(_misMatchedBootConfig);
			
			//Act
			Assert.DoesNotThrowAsync(async () => await client.StartAsync());
			
			//Assert
			Assert.AreEqual(1,client.ShardClients.Count);
		}

		[Test]
		public async Task ClusterClient_Connects_With_Specified_ShardId_When_Specified()
		{
			//Arrange
			var conf1 = new DiscordConfiguration(_bootShardConfig) {ShardCount = 4, ClusterId = 0};
			var conf2 = new DiscordConfiguration(conf1) {ClusterId = 1};

			var client1 = new DiscordClusterClient(conf1);
			var client2 = new DiscordClusterClient(conf2);
			
			//Act
			Assert.DoesNotThrowAsync(async () => await client1.StartAsync());
			Assert.DoesNotThrowAsync(async () => await client2.StartAsync());
			
			//Assert
			Assert.AreEqual(0, client1.ShardClients[0].ShardId);
			Assert.AreEqual(2, client2.ShardClients[0].ShardId);
		}
		
		[Test]
		public async Task ClusterClient_OnlyConnects_Specified_Shards()
		{
			//Arrange
			var conf1 = new DiscordConfiguration(_bootShardConfig) {ShardCount = 4, ClusterId = 0};
			var conf2 = new DiscordConfiguration(conf1) {ClusterId = 1};

			var client1 = new DiscordClusterClient(conf1);
			var client2 = new DiscordClusterClient(conf2);
			
			//Act
			Assert.DoesNotThrowAsync(async () => await client1.StartAsync());
			Assert.DoesNotThrowAsync(async () => await client2.StartAsync());
			
			//Assert
			Assert.AreEqual(2, client1.ShardClients.Count);
			Assert.AreEqual(2, client2.ShardClients.Count);
		}

		[Test]
		public void DiscordConfiguration_Disallows_Invalid_NonNull_Values()
		{
			//Arrange
			var config = new DiscordConfiguration();
			
			//Act
			//Assert
			Assert.Throws<ArgumentOutOfRangeException>(() => config.BootShardCount = 0);
			Assert.Throws<ArgumentOutOfRangeException>(() => config.BootShardCount = 2);
		}

		[Test]
		public void DiscordConfiguration_Disallows_Invalid_Clustering()
		{
			//Arrange
			var config = new DiscordConfiguration() {ShardCount = 2, BootShardCount = 2};
			
			//Act
			//Assert
			Assert.Throws<ArgumentOutOfRangeException>(() => config.ClusterId = -1);
			Assert.Throws<ArgumentOutOfRangeException>(() => config.ClusterId = 1);
		}
	}
}
```

**Comment by VelvetThePanda** [comment]
I will do some testing to ensure clusters ensure shards return the correct shard Id as this is issue @uwx brought up and I forgot to mention.

File: DSharpPlus/Logging/DefaultLoggerFactory.cs [hunk 1]
```diff
@@ -32,14 +33,21 @@ internal class DefaultLoggerFactory : ILoggerFactory
         private List<ILoggerProvider> Providers { get; } = new List<ILoggerProvider>();
         private bool _isDisposed = false;
 
+        private readonly string[] _supoprtedLoggerTypes = new[]
```
Comment (uwx) [inline DSharpPlus/Logging/DefaultLoggerFactory.cs:36]: Grammar.

File: DSharpPlus/DiscordConfiguration.cs [hunk 2]
```diff
@@ -100,6 +101,46 @@ public string Token
...
+            set
+            {
+                if ((value != null && value > this.ShardCount) || value < 1)
+                    throw new ArgumentOutOfRangeException(nameof(value), "Value must be greater than zero and less than or equal to ShardCount.");
+
+                this._shardBootCount = value;
+            }
+        }
+
+        private int? _shardBootCount;
```
Comment (uwx) [inline DSharpPlus/DiscordConfiguration.cs:120]: Inconsistent naming.

File: DSharpPlus/Logging/ShardedLoggerFactory.cs [hunk 3]
```diff
@@ -26,23 +26,18 @@
 
 namespace DSharpPlus
 {
-    internal class ShardedLoggerFactory : ILoggerFactory
+    internal class ShardedLoggerFactory<T> : ILoggerFactory
```
Comment (uwx) [inline DSharpPlus/Logging/ShardedLoggerFactory.cs:29]: Consider `where T : BaseDiscordClient`, as that seems to be the only appropriate use of this class.

File: DSharpPlus/Logging/ShardedLoggerFactory.cs [hunk 4]
```diff
@@ -26,23 +26,18 @@
...
 
         public void AddProvider(ILoggerProvider provider) => throw new InvalidOperationException("This is a passthrough logger container, it cannot register new providers.");
 
-        public ILogger CreateLogger(string categoryName)
-        {
-            return categoryName != typeof(BaseDiscordClient).FullName
-                ? throw new ArgumentException($"This factory can only provide instances of loggers for {typeof(BaseDiscordClient).FullName}.", nameof(categoryName))
-                : this.Logger;
-        }
+        public ILogger CreateLogger(string categoryName) => this.Logger;
```
Comment (uwx) [inline DSharpPlus/Logging/ShardedLoggerFactory.cs:40]: DefaultLoggerFactory has category name checks, shouldn't they be here too?

**Review by uwx — CHANGES_REQUESTED** [review]

File: DSharpPlus/Logging/ShardedLoggerFactory.cs (hunk 3 above)
Comment (VelvetThePanda) [reply DSharpPlus/Logging/ShardedLoggerFactory.cs:29]: Sharded nor Clustered client extend the BaseDiscordClient, though

**Comment by Neuheit** [comment]
Closing due to discussion in the Discord.

**Pull request closed without merging**